Tests/AWSXRayInstrumentTests/InstrumentTests.swift

```diff
@@ -0,0 +1,32 @@
+import XCTest
+
+import AWSXRayInstrument
+import AWSXRayRecorder
+import Baggage
+import Dispatch
+import Instrumentation
+
+final class InstrumentTests: XCTestCase {
+    func testRecordingOneSegment() {
+        let emitter = TestEmitter()
+        let instrument: TracingInstrument = XRayRecorder(emitter: emitter)
+
+        XCTAssertEqual(0, emitter.segments.count)
+
+        let name: String = UUID().uuidString
+        let context = BaggageContext()
+        let kind = SpanKind.internal
+        let now = DispatchTime.now()
+
+        var span: Span = instrument.startSpan(named: name, context: context, ofKind: kind, at: now)
+
+        // TODO: test attributes
+
+        span.end()
+
+        // TODO: instrument does not define any method to flush/wait
+        (instrument as? XRayRecorder)?.wait()
+
```

Comment:
@slashmo not sure it should be exposed, but thats what arguably every instrument will need to do

in case of containers this will probably need to be called just before program exists,
in case of Lambda - events need to be flushed while the lambda is running

Reply:
Could you handle this in the recorder's shutdown?

We'd offer a `shutdown(...)` on the tracing SPI which is the general thing we're after, rather than exposing a flush;

Generally following the spec:


> ### Shutdown and explicit flushing could block
> 
> The language library could block the end-user application when it shut down. On shutdown, it has to flush data to prevent information loss. The language library should support user-configurable timeout if it blocks on shut down.
> 
> If the language library supports an explicit flush operation, it could block also. But should support a configurable timeout.

Reply:
@ktoso

> Could you handle this in the recorder's shutdown?

yes for "services" (Vapor, Kitara, ...)

however in case of "lambda runtime", my understanding is that we do need to emit all the segments during event handling (event though it will make the request handling longer) because otherwise lambda may be suspended (and potentially never wake up again) before emitting finishes, example:

```swift
private struct ExampleLambdaHandler: EventLoopLambdaHandler {
    typealias In = Cloudwatch.ScheduledEvent
    typealias Out = Void

    private let recorder = XRayRecorder()

    private func doWork(on eventLoop: EventLoop) -> EventLoopFuture<Void> {
        eventLoop.submit { usleep(100_000) }.map { _ in }
    }

    func handle(context: Lambda.Context, event: In) -> EventLoopFuture<Void> {
        recorder.segment(name: "ExampleLambdaHandler", context: context) {
            self.doWork(on: context.eventLoop)
        }.flatMap {
            self.recorder.flush(on: context.eventLoop)
        }
    }
}

Lambda.run(ExampleLambdaHandler())
```

to be honest, while my explanation above does make sense to me, I do plan checking more in details how its implemented/enforced in "official" XRay SDKs;

also note that while in case of XRay emitting = sending UDP in local network, which should be comparatively inexpensive and may be finished (subject of more testing) before lambda is suspended even without explicit flushing;
emitting spans using different `TracingInstrument` in lambda handler will be typically much mroe expensive and will take more time;

Reply:
let's follow up on https://github.com/slashmo/gsoc-swift-tracing/issues/85